packages/sanes-chrome-extension/src/extension/bsMessageHandler/externalHandler.ts

```diff
@@ -0,0 +1,62 @@
+/*global chrome*/
+import { JsonRpcErrorResponse, JsonRpcRequest } from '@iov/jsonrpc';
+import { UseOnlyJsonRpcSigningServer } from '../../logic/persona';
+
+type SigningServer = UseOnlyJsonRpcSigningServer | undefined;
+type SendResponse = (response: any) => void; // eslint-disable-line @typescript-eslint/no-explicit-any
+
+function handleWithServer(
+  signingServer: SigningServer,
+  request: JsonRpcRequest,
+  sendResponse: SendResponse
+): void {
+  if (!signingServer) {
+    console.warn('Could not pass message to signing server');
+    const error: JsonRpcErrorResponse = {
+      jsonrpc: '2.0',
+      id: typeof request.id === 'number' ? request.id : null,
+      error: {
+        code: -32000,
+        message: 'Signing server not ready',
+      },
+    };
+    sendResponse(error);
+    return;
+  }
+
+  signingServer
+    .handleUnchecked(request)
+    .then(response => {
+      console.log('Responding', response);
+      sendResponse(response);
+    })
+    .catch(console.error);
+}
+
+function isForServer(request: JsonRpcRequest): boolean {
+  const { method } = request;
+  return method === 'getIdentities' || method === 'signAndPost';
+}
+
+function handleResponseToExtension(request: JsonRpcRequest, sendResponse: SendResponse): void {
+  chrome.browserAction.setIcon({ path: 'assets/icons/request128.png' });
+  chrome.browserAction.setBadgeBackgroundColor({ color: [0, 0, 0, 0] });
+  chrome.browserAction.setBadgeText({ text: 'REQ' });
+
+  // TODO dispath info to redux for getting the info in the extension. Include the
+  // sendResponse ref for handling answer
+  sendResponse(`I have handled the ${request.method} request changing the icon`);
+}
+
+export function handleExternalMessage(
```

Comment:
This function could be converted into an async function returning a promise (like `handleInternalMessage`) to avoid all the callback hell

Reply:
Also for better handling of errors. 👌